openedx_learning/apps/authoring/linking/models.py

```diff
@@ -0,0 +1,132 @@
+]
+
+
+class PublishableEntityLink(models.Model):
+    """
+    This represents link between any two publishable entities or link between publishable entity and a course
+    xblock. It helps in tracking relationship between xblocks imported from libraries and used in different courses.
+    """
+    uuid = immutable_uuid_field()
+    upstream_block = models.ForeignKey(
+        PublishableEntity,
+        on_delete=models.SET_NULL,
+        related_name="links",
+        null=True,
+        blank=True,
+    )
+    upstream_usage_key = key_field(
+        help_text=_(
+            "Upstream block usage key, this value cannot be null"
+            " and useful to track upstream library blocks that do not exist yet"
+        )
+    )
+    upstream_context_key = case_sensitive_char_field(
+        max_length=500,
+        help_text=_(
+            "Upstream context key i.e., learning_package/library key"
+        ),
+        null=True,
+        blank=True,
+    )
+    downstream_usage_key = key_field()
+    downstream_context_key = key_field()
+    downstream_context_title = case_insensitive_char_field(
+        null=False,
+        blank=False,
+        max_length=1000,
+        help_text=_(
+            "The title of downstream context, for example, course display name."
+        ),
+    )
+    version_synced = models.IntegerField()
+    version_declined = models.IntegerField(null=True, blank=True)
+    created = manual_date_time_field()
+    updated = manual_date_time_field()
+
+    def __str__(self):
+        return f"{self.upstream_usage_key}->{self.downstream_usage_key}"
+
+    class Meta:
+        constraints = [
+            # A downstream entity can only link to single upstream entity
+            # whereas an entity can be upstream for multiple downstream entities.
+            models.UniqueConstraint(
+                fields=["downstream_usage_key"],
+                name="oel_link_ent_downstream_key",
+            )
+        ]
+        indexes = [
+            # Search by course/downstream key
+            models.Index(
+                fields=["downstream_context_key"],
+                name="oel_link_ent_idx_down_ctx_key",
+            ),
+            # Search by library/upstream key
+            models.Index(
+                fields=["upstream_context_key"],
+                name="oel_link_ent_idx_up_ctx_key"
+            ),
+        ]
+        verbose_name = "Publishable Entity Link"
+        verbose_name_plural = "Publishable Entity Links"
+
+
+class CourseLinksStatusChoices(models.TextChoices):
+    """
+    Enumerates the states that a CourseLinksStatus can be in.
+    """
+    PENDING = "pending", _("Pending")
+    PROCESSING = "processing", _("Processing")
+    FAILED = "failed", _("Failed")
+    COMPLETED = "completed", _("Completed")
+
+
+class CourseLinksStatus(models.Model):
+    """
+    This table stores current processing status of upstream-downstream links in PublishableEntityLink table for a
+    course.
```

Comment:
Is there ever a case where we'd want to re-process the downstream links for a given library (and therefore want to track them too)? I ask only because the naming for this model is very course-specific, but the data itself doesn't care about courses -- any learning context should work fine here.

Reply:
Dave [suggested](https://docs.google.com/document/d/168F2VNz7sy1uX8RhblvVl0WLTl8yMUlXANY2LISFFWQ/edit?disco=AAABaR9eTts) tracking these courses so that we can trigger processing when a course author visits sync page of a course, to avoid processing all of them in a migration.

Reply:
> any learning context should work fine here.

True, we should probably rename the table.